File: javascript/validador_cpf.js
function gerarDigitoVerificador(cpf) {
  let soma1 = 0;
  for (let i = 0; i < 9; i++) {
    soma1 += parseInt(cpf[i]) * (10 - i);
  }

  let digito1 = soma1 % 11;
  if (digito1 === 0 || digito1 === 1) {
    digito1 = 0;
  } else {
    digito1 = 11 - digito1;
  }

  let soma2 = 0;
  for (let i = 0; i < 9; i++) {
    soma2 += parseInt(cpf[i]) * (11 - i);
  }
  soma2 += digito1 * 2;

  let digito2 = soma2 % 11;
  if (digito2 === 0 || digito2 === 1) {
    digito2 = 0;
  } else {
    digito2 = 11 - digito2;
  }

  console.log(
    `O CPF completo é: ${cpf.slice(0, 3)}.${cpf.slice(3, 6)}.${cpf.slice(6, 9)}-${digito1}${digito2}`
  );
  return `${digito1}${digito2}`;
}

function verificarCpf(cpf) {
  let digitoVerificador1 = parseInt(cpf[9]);
  let digitoVerificador2 = parseInt(cpf[10]);
  cpf = cpf.slice(0, 9);

  let soma1 = 0;
  for (let i = 0; i < 9; i++) {
    soma1 += parseInt(cpf[i]) * (10 - i);
  }

  let digito1 = soma1 % 11;
  if (digito1 === 0 || digito1 === 1) {
    digito1 = 0;
  } else {
    digito1 = 11 - digito1;
  }

  let soma2 = 0;
  for (let i = 0; i < 9; i++) {
    soma2 += parseInt(cpf[i]) * (11 - i);
  }
  soma2 += digito1 * 2;

  let digito2 = soma2 % 11;
  if (digito2 === 0 || digito2 === 1) {
    digito2 = 0;
  } else {
    digito2 = 11 - digito2;
  }

  if (digito1 === digitoVerificador1 && digito2 === digitoVerificador2) {
    console.log(
      `O CPF: ${cpf.slice(0, 3)}.${cpf.slice(3, 6)}.${cpf.slice(6, 9)}-${digito1}${digito2} é válido`
    );
    return "Válido";
  } else {
    console.log(
      `O CPF: ${cpf.slice(0, 3)}.${cpf.slice(3, 6)}.${cpf.slice(6, 9)}-${digito1}${digito2} não é válido`
    );
    return "Inválido";
  }
}

function validadorCpf(cpf) {
  // Remove tudo que não é número
  cpf = cpf.replace(/\D/g, "");

  let numeroDigitos = cpf.length;

  if (numeroDigitos < 9 || numeroDigitos > 11) {
    return "Inválido";
  }

  if (numeroDigitos === 9) {
    return gerarDigitoVerificador(cpf);
  }

  if (numeroDigitos === 11) {
    return verificarCpf(cpf);
  }
}

// Programa principal
const entradaUsuario = prompt("Entre com o número do CPF para validar:");
const resultado = validadorCpf(entradaUsuario);
console.log(resultado);

File: python/validador_cpf.py
import re


def _gerador_digito_verificador(cpf: int) -> str:
    soma1 = sum(int(cpf[i]) * (10-i) for i in range(9))
    digito1 = soma1 % 11

    if digito1 == 0 or digito1 == 1:
        digito1 = 0
    else:
        digito1 = 11 - digito1

    soma2 = sum(int(cpf[i]) * (11-i) for i in range(9)) + digito1 * 2

    digito2 = soma2 % 11

    if digito2 == 0 or digito2 == 1:
        digito2 = 0
    else:
        digito2 = 11 - digito2

    print(f"O CPF completo é: {cpf[:3]}.{cpf[3:6]}.{cpf[6:9]}-{digito1}{digito2}")
    return str(f"{digito1}{digito2}")


def _verificador_cpf(cpf):

    digito_verificador1, digito_verificador2 = int(cpf[9]), int(cpf[10])
    cpf = cpf[:9]
    
    soma1 = sum(int(cpf[i]) * (10-i) for i in range(9))
    digito1 = soma1 % 11

    if digito1 == 0 or digito1 == 1:
        digito1 = 0
    else:
        digito1 = 11 - digito1

    soma2 = sum(int(cpf[i]) * (11-i) for i in range(9)) + digito1 * 2

    digito2 = soma2 % 11

    if digito2 == 0 or digito2 == 1:
        digito2 = 0
    else:
        digito2 = 11 - digito2

    if digito1 == digito_verificador1 and digito2 == digito_verificador2:
        print(f"O CPF: {cpf[:3]}.{cpf[3:6]}.{cpf[6:9]}-{digito1}{digito2} é válido")
        return "Válido"
    else:
        print(f"O CPF: {cpf[:3]}.{cpf[3:6]}.{cpf[6:9]}-{digito1}{digito2} não é válido")
        return "Inválido"




def validador_cpf(cpf: str) -> str:

    cpf = re.sub(r'\D', '', cpf)

    
    numero_digitos = len(cpf)

    if numero_digitos < 9 or numero_digitos > 11:
        return 'Inválido'

    if numero_digitos == 9:
       return _gerador_digito_verificador(cpf)

    if numero_digitos == 11:
        return _verificador_cpf(cpf)
    


if __name__ == '__main__':
    entrada_usuario = input("Entre com o numero do CPF para validar: ")
    resultado = validador_cpf(entrada_usuario)
    print(resultado)